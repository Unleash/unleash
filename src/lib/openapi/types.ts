import { OpenAPIV3 } from 'openapi-types';
import { FromSchema } from 'json-schema-to-ts';
import { DeepMutable } from '../types/mutable';

// Admin paths must have the "admin" tag.
export interface AdminApiOperation
    extends Omit<OpenAPIV3.OperationObject, 'tags'> {
    tags: ['admin'];
}

// Client paths must have the "client" tag.
export interface ClientApiOperation
    extends Omit<OpenAPIV3.OperationObject, 'tags'> {
    tags: ['client'];
}

// Create a type from a const schema object.
export type CreateSchemaType<T> = FromSchema<T>;

// Create an OpenAPIV3.SchemaObject from a const schema object.
export const createSchemaObject = <T>(schema: T): DeepMutable<T> => schema;
