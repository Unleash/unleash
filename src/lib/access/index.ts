import { Db, IUnleashConfig } from 'lib/server-impl';
import EventStore from '../db/event-store';
import GroupStore from '../db/group-store';
import { AccountStore } from '../db/account-store';
import RoleStore from '../db/role-store';
import EnvironmentStore from '../db/environment-store';
import { AccessStore } from '../db/access-store';
import { AccessService, GroupService } from '../services';

export const createAccessService = (
    db: Db,
    config: IUnleashConfig,
): AccessService => {
    const { eventBus, getLogger } = config;
    const eventStore = new EventStore(db, getLogger);
    const groupStore = new GroupStore(db);
    const accountStore = new AccountStore(db, getLogger);
    const roleStore = new RoleStore(db, eventBus, getLogger);
    const environmentStore = new EnvironmentStore(db, eventBus, getLogger);
    const accessStore = new AccessStore(db, eventBus, getLogger);
    const groupService = new GroupService(
        { groupStore, eventStore, accountStore },
        { getLogger },
    );

    return new AccessService(
        { accessStore, accountStore, roleStore, environmentStore },
        { getLogger },
        groupService,
    );
};
