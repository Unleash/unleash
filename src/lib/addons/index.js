const webhook = require('./webhook');
const slackAddon = require('./slack');
const teamsAddon = require('./teams');
const datadogAddon = require('./datadog');

const addons = [webhook, slackAddon, teamsAddon, datadogAddon];

module.exports = addons;
