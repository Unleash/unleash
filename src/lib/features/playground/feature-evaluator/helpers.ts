import { IStrategyConfig } from '../../../types';
import { FeatureStrategiesEvaluationResult } from './client';
import { Context } from './context';

export type FallbackFunction = (name: string, context: Context) => boolean;

export function createFallbackFunction(
    name: string,
    context: Context,
    fallback?: FallbackFunction | boolean,
): () => FeatureStrategiesEvaluationResult {
    const createEvalResult = (enabled: boolean) => ({
        result: enabled,
        strategies: [],
    });

    if (typeof fallback === 'function') {
        return () => createEvalResult(fallback(name, context));
    }
    if (typeof fallback === 'boolean') {
        return () => createEvalResult(fallback);
    }
    return () => createEvalResult(false);
}

export function resolveContextValue(
    context: Context,
    field: string,
): string | undefined {
    if (context[field]) {
        return context[field] as string;
    }
    if (context.properties?.[field]) {
        return context.properties[field] as string;
    }
    return undefined;
}

export function safeName(str: string = ''): string {
    return str.replace(/\//g, '_');
}

export function getDefaultStrategy(featureName: string): IStrategyConfig {
    return {
        name: 'flexibleRollout',
        constraints: [],
        disabled: false,
        parameters: {
            rollout: '100',
            stickiness: 'default',
            groupId: featureName,
        },
    };
}

function resolveGroupId(
    defaultStrategy: IStrategyConfig,
    featureName: string,
): string {
    const groupId =
        defaultStrategy?.parameters?.groupId !== ''
            ? defaultStrategy.parameters?.groupId
            : featureName;

    return groupId || '';
}

export function getProjectDefaultStrategy(
    defaultStrategy: IStrategyConfig,
    featureName: string,
): IStrategyConfig {
    return {
        ...defaultStrategy,
        parameters: {
            ...defaultStrategy.parameters,
            groupId: resolveGroupId(defaultStrategy, featureName),
        },
    };
}
