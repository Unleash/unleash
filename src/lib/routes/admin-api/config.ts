import { Response } from 'express';
import { AuthedRequest } from '../../types/core';
import { IUnleashServices } from '../../types/services';
import { IAuthType, IUnleashConfig } from '../../types/option';
import version from '../../util/version';
import Controller from '../controller';
import VersionService from '../../services/version-service';
import SettingService from '../../services/setting-service';
import {
    simpleAuthSettingsKey,
    SimpleAuthSettings,
} from '../../types/settings/simple-auth-settings';
import { ADMIN, NONE } from '../../types/permissions';
import { createResponseSchema } from '../../openapi/util/create-response-schema';
import {
    uiConfigSchema,
    UiConfigSchema,
} from '../../openapi/spec/ui-config-schema';
import { OpenApiService } from '../../services/openapi-service';
import { EmailService } from '../../services/email-service';
import { emptyResponse } from '../../openapi/util/standard-responses';
import { IAuthRequest } from '../unleash-types';
import { extractUsername } from '../../util/extract-user';
import NotFoundError from '../../error/notfound-error';
import { SetUiConfigSchema } from '../../openapi/spec/set-ui-config-schema';
import { createRequestSchema } from '../../openapi/util/create-request-schema';

class ConfigController extends Controller {
    private versionService: VersionService;

    private settingService: SettingService;

    private emailService: EmailService;

    private readonly openApiService: OpenApiService;

    constructor(
        config: IUnleashConfig,
        {
            versionService,
            settingService,
            emailService,
            openApiService,
        }: Pick<
            IUnleashServices,
            | 'versionService'
            | 'settingService'
            | 'emailService'
            | 'openApiService'
        >,
    ) {
        super(config);
        this.versionService = versionService;
        this.settingService = settingService;
        this.emailService = emailService;
        this.openApiService = openApiService;

        this.route({
            method: 'get',
            path: '',
            handler: this.getUiConfig,
            permission: NONE,
            middleware: [
                openApiService.validPath({
                    tags: ['Admin UI'],
                    operationId: 'getUiConfig',
                    responses: {
                        200: createResponseSchema('uiConfigSchema'),
                    },
                }),
            ],
        });

        this.route({
            method: 'post',
            path: '',
            handler: this.setUiConfig,
            permission: ADMIN,
            middleware: [
                openApiService.validPath({
                    tags: ['Admin UI'],
                    operationId: 'setUiConfig',
                    requestBody: createRequestSchema('setUiConfigSchema'),
                    responses: { 200: emptyResponse },
                }),
            ],
        });
    }

    async getUiConfig(
        req: AuthedRequest,
        res: Response<UiConfigSchema>,
    ): Promise<void> {
        const [frontendSettings, simpleAuthSettings] = await Promise.all([
            this.settingService.getFrontendSettings(),
            this.settingService.get<SimpleAuthSettings>(simpleAuthSettingsKey),
        ]);

        const disablePasswordAuth =
            simpleAuthSettings?.disabled ||
            this.config.authentication.type == IAuthType.NONE;

        const expFlags = this.config.flagResolver.getAll({
            email: req.user.email,
        });
        const flags = { ...this.config.ui.flags, ...expFlags };

        const response: UiConfigSchema = {
            ...this.config.ui,
            flags,
            version,
            emailEnabled: this.emailService.isEnabled(),
            unleashUrl: this.config.server.unleashUrl,
            baseUriPath: this.config.server.baseUriPath,
            authenticationType: this.config.authentication?.type,
            segmentValuesLimit: this.config.segmentValuesLimit,
            strategySegmentsLimit: this.config.strategySegmentsLimit,
            frontendApiOrigins: frontendSettings.frontendApiOrigins,
            versionInfo: this.versionService.getVersionInfo(),
            disablePasswordAuth,
        };

        this.openApiService.respondWithValidation(
            200,
            res,
            uiConfigSchema.$id,
            response,
        );
    }

    async setUiConfig(
        req: IAuthRequest<void, void, SetUiConfigSchema>,
        res: Response<string>,
    ): Promise<void> {
        if (req.body.frontendSettings) {
            await this.settingService.setFrontendSettings(
                req.body.frontendSettings,
                extractUsername(req),
            );
            res.sendStatus(204);
            return;
        }

        throw new NotFoundError();
    }
}

export default ConfigController;
