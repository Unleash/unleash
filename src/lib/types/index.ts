export * from './services';
export * from './stores';
export * from './option';
