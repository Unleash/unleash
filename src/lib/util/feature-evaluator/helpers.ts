import { IStrategyConfig } from '../../types/model';
import { FeatureStrategiesEvaluationResult } from './client';
import { Context } from './context';

export type FallbackFunction = (name: string, context: Context) => boolean;

export function createFallbackFunction(
    name: string,
    context: Context,
    fallback?: FallbackFunction | boolean,
): () => FeatureStrategiesEvaluationResult {
    const createEvalResult = (enabled: boolean) => ({
        result: enabled,
        strategies: [],
    });

    if (typeof fallback === 'function') {
        return () => createEvalResult(fallback(name, context));
    }
    if (typeof fallback === 'boolean') {
        return () => createEvalResult(fallback);
    }
    return () => createEvalResult(false);
}

export function resolveContextValue(
    context: Context,
    field: string,
): string | undefined {
    if (context[field]) {
        return context[field] as string;
    }
    if (context.properties && context.properties[field]) {
        return context.properties[field] as string;
    }
    return undefined;
}

export function safeName(str: string = ''): string {
    return str.replace(/\//g, '_');
}

export function getDefaultStrategy(featureName: string): IStrategyConfig {
    return {
        name: 'flexibleRollout',
        constraints: [],
        parameters: {
            rollout: '100',
            stickiness: 'default',
            groupId: featureName,
        },
    };
}
