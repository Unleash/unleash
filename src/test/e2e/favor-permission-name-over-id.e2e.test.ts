import { getDbConfig } from './helpers/database-config';
import { createTestConfig } from '../config/test-config';
import { getInstance } from 'db-migrate';
import { log } from 'db-migrate-shared';
import { Client } from 'pg';
import { IDBOption } from '../../lib/types';

log.setLogLevel('error');

async function initSchema(db: IDBOption): Promise<void> {
    const client = new Client(db);
    await client.connect();
    await client.query(`DROP SCHEMA IF EXISTS ${db.schema} CASCADE`);
    await client.query(`CREATE SCHEMA IF NOT EXISTS ${db.schema}`);
    await client.end();
}

test('Favor permission name over id migration correctly assigns permissions by name', async () => {
    jest.setTimeout(15000);
    const config = createTestConfig({
        db: {
            ...getDbConfig(),
            pool: { min: 1, max: 4 },
            schema: 'favor_permission_name_over_id_test',
            ssl: false,
        },
    });

    await initSchema(config.db);

    const custom = {
        ...config.db,
        connectionTimeoutMillis: 2000,
    };

    // disable Intellij/WebStorm from setting verbose CLI argument to db-migrator
    process.argv = process.argv.filter((it) => !it.includes('--verbose'));
    const dbm = getInstance(true, {
        cwd: `${__dirname}/../../`, // relative to src/test/e2e
        config: { custom },
        env: 'custom',
    });

    dbm.config.custom = custom;

    // Run all migrations up to, and including, this one, the last one before the favor permission name over id migration
    await dbm.up('20231123100052-drop-last-seen-foreign-key.js');

    // Set up the test data
    const client = new Client(config.db);
    await client.connect();
    await client.query(`
        DELETE FROM "favor_permission_name_over_id_test"."roles";

        INSERT INTO "favor_permission_name_over_id_test"."roles" (id, name, description, type) VALUES (101, 'Role 1', 'A test role', 'custom');
    `);
    await client.query(`
        DELETE FROM "favor_permission_name_over_id_test"."permissions";

        INSERT INTO "favor_permission_name_over_id_test"."permissions" (id, permission, display_name, type) VALUES (101, 'TEST_PERMISSION_1', 'A test permission', 'root');
        INSERT INTO "favor_permission_name_over_id_test"."permissions" (id, permission, display_name, type) VALUES (102, 'TEST_PERMISSION_2', 'A test permission', 'root');
        INSERT INTO "favor_permission_name_over_id_test"."permissions" (id, permission, display_name, type) VALUES (103, 'TEST_PERMISSION_3', 'A test permission', 'root');
        INSERT INTO "favor_permission_name_over_id_test"."permissions" (id, permission, display_name, type) VALUES (104, 'TEST_PERMISSION_4', 'A test permission', 'root');
    `);
    await client.query(`
        DELETE FROM "favor_permission_name_over_id_test"."role_permission";

        INSERT INTO "favor_permission_name_over_id_test"."role_permission" (role_id, permission_id) VALUES (101, 101);
        INSERT INTO "favor_permission_name_over_id_test"."role_permission" (role_id, permission_id) VALUES (101, 102);
        INSERT INTO "favor_permission_name_over_id_test"."role_permission" (role_id, permission_id) VALUES (101, 103);
        INSERT INTO "favor_permission_name_over_id_test"."role_permission" (role_id, permission_id) VALUES (101, 104);
    `);

    // Run the drop permissions id migration
    await dbm.up('20231123155649-favor-permission-name-over-id.js');

    // Check the results
    const { rows: resultsPermissions } = await client.query(`
        SELECT * FROM "favor_permission_name_over_id_test"."permissions" ORDER BY created_at;
    `);

    const { rows: resultsRolePermission } = await client.query(`
        SELECT * FROM "favor_permission_name_over_id_test"."role_permission" WHERE role_id = 101 ORDER BY created_at;
    `);

    // We keep the id for now, but will remove it in the next major version
    expect('id' in resultsPermissions[0]).toEqual(true);
    expect(
        resultsPermissions.map(({ permission, display_name, type }) => ({
            permission,
            display_name,
            type,
        })),
    ).toMatchSnapshot();

    // We keep the permission_id for now, but will remove it in the next minor version
    expect('permission_id' in resultsRolePermission[0]).toEqual(true);
    expect('permission' in resultsRolePermission[0]).toEqual(true);

    expect(resultsRolePermission.length).toEqual(4);
    expect(
        resultsRolePermission.map(({ role_id, permission }) => ({
            role_id,
            permission,
        })),
    ).toEqual([
        { role_id: 101, permission: 'TEST_PERMISSION_1' },
        { role_id: 101, permission: 'TEST_PERMISSION_2' },
        { role_id: 101, permission: 'TEST_PERMISSION_3' },
        { role_id: 101, permission: 'TEST_PERMISSION_4' },
    ]);

    // Check the results that ensure the default roles exist and have the correct permissions
    const { rows: resultsRoles } = await client.query(`
        SELECT name, description, type FROM "favor_permission_name_over_id_test"."roles" WHERE id != 101 ORDER BY created_at;
    `);

    const { rows: resultsDefaultRolePermissions } = await client.query(`
        SELECT role_id, environment, permission FROM "favor_permission_name_over_id_test"."role_permission" WHERE role_id != 101 ORDER BY created_at;
    `);

    expect(resultsRoles.length).toEqual(5);
    expect(resultsRoles).toEqual([
        {
            name: 'Admin',
            description:
                'Users with the root admin role have superuser access to Unleash and can perform any operation within the Unleash platform.',
            type: 'root',
        },
        {
            name: 'Editor',
            description:
                'Users with the root editor role have access to most features in Unleash, but can not manage users and roles in the root scope. Editors will be added as project owners when creating projects and get superuser rights within the context of these projects. Users with the editor role will also get access to most permissions on the default project by default.',
            type: 'root',
        },
        {
            name: 'Viewer',
            description:
                'Users with the root viewer role can only read root resources in Unleash. Viewers can be added to specific projects as project members. Users with the viewer role may not view API tokens.',
            type: 'root',
        },
        {
            name: 'Owner',
            description:
                'Users with the project owner role have full control over the project, and can add and manage other users within the project context, manage feature toggles within the project, and control advanced project features like archiving and deleting the project.',
            type: 'project',
        },
        {
            name: 'Member',
            description:
                "Users with the project member role are allowed to view, create, and update feature toggles within a project, but have limited permissions in regards to managing the project's user access and can not archive or delete the project.",
            type: 'project',
        },
    ]);

    expect(resultsDefaultRolePermissions).toMatchSnapshot();

    await client.end();
    await dbm.reset();
});
