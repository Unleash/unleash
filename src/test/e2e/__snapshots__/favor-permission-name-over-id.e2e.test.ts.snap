// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favor permission name over id migration correctly assigns permissions by name 1`] = `
[
  {
    "display_name": "A test permission",
    "permission": "TEST_PERMISSION_1",
    "type": "root",
  },
  {
    "display_name": "A test permission",
    "permission": "TEST_PERMISSION_2",
    "type": "root",
  },
  {
    "display_name": "A test permission",
    "permission": "TEST_PERMISSION_3",
    "type": "root",
  },
  {
    "display_name": "A test permission",
    "permission": "TEST_PERMISSION_4",
    "type": "root",
  },
  {
    "display_name": "Admin",
    "permission": "ADMIN",
    "type": "root",
  },
  {
    "display_name": "Create feature toggles",
    "permission": "CREATE_FEATURE",
    "type": "project",
  },
  {
    "display_name": "Create activation strategies",
    "permission": "CREATE_STRATEGY",
    "type": "root",
  },
  {
    "display_name": "Create addons",
    "permission": "CREATE_ADDON",
    "type": "root",
  },
  {
    "display_name": "Delete addons",
    "permission": "DELETE_ADDON",
    "type": "root",
  },
  {
    "display_name": "Update addons",
    "permission": "UPDATE_ADDON",
    "type": "root",
  },
  {
    "display_name": "Update feature toggles",
    "permission": "UPDATE_FEATURE",
    "type": "project",
  },
  {
    "display_name": "Delete feature toggles",
    "permission": "DELETE_FEATURE",
    "type": "project",
  },
  {
    "display_name": "Update applications",
    "permission": "UPDATE_APPLICATION",
    "type": "root",
  },
  {
    "display_name": "Update tag types",
    "permission": "UPDATE_TAG_TYPE",
    "type": "root",
  },
  {
    "display_name": "Delete tag types",
    "permission": "DELETE_TAG_TYPE",
    "type": "root",
  },
  {
    "display_name": "Create projects",
    "permission": "CREATE_PROJECT",
    "type": "root",
  },
  {
    "display_name": "Update project",
    "permission": "UPDATE_PROJECT",
    "type": "project",
  },
  {
    "display_name": "Delete project",
    "permission": "DELETE_PROJECT",
    "type": "project",
  },
  {
    "display_name": "Update strategies",
    "permission": "UPDATE_STRATEGY",
    "type": "root",
  },
  {
    "display_name": "Delete strategies",
    "permission": "DELETE_STRATEGY",
    "type": "root",
  },
  {
    "display_name": "Update context fields",
    "permission": "UPDATE_CONTEXT_FIELD",
    "type": "root",
  },
  {
    "display_name": "Create context fields",
    "permission": "CREATE_CONTEXT_FIELD",
    "type": "root",
  },
  {
    "display_name": "Delete context fields",
    "permission": "DELETE_CONTEXT_FIELD",
    "type": "root",
  },
  {
    "display_name": "Read roles",
    "permission": "READ_ROLE",
    "type": "root",
  },
  {
    "display_name": "Create activation strategies",
    "permission": "CREATE_FEATURE_STRATEGY",
    "type": "environment",
  },
  {
    "display_name": "Update activation strategies",
    "permission": "UPDATE_FEATURE_STRATEGY",
    "type": "environment",
  },
  {
    "display_name": "Delete activation strategies",
    "permission": "DELETE_FEATURE_STRATEGY",
    "type": "environment",
  },
  {
    "display_name": "Create CLIENT API tokens",
    "permission": "CREATE_CLIENT_API_TOKEN",
    "type": "root",
  },
  {
    "display_name": "Create/edit variants",
    "permission": "UPDATE_FEATURE_VARIANTS",
    "type": "project",
  },
  {
    "display_name": "Change feature toggle project",
    "permission": "MOVE_FEATURE_TOGGLE",
    "type": "project",
  },
  {
    "display_name": "Create segments",
    "permission": "CREATE_SEGMENT",
    "type": "root",
  },
  {
    "display_name": "Edit segments",
    "permission": "UPDATE_SEGMENT",
    "type": "root",
  },
  {
    "display_name": "Delete segments",
    "permission": "DELETE_SEGMENT",
    "type": "root",
  },
  {
    "display_name": "Read api tokens for a specific project",
    "permission": "READ_PROJECT_API_TOKEN",
    "type": "project",
  },
  {
    "display_name": "Create api tokens for a specific project",
    "permission": "CREATE_PROJECT_API_TOKEN",
    "type": "project",
  },
  {
    "display_name": "Delete api tokens for a specific project",
    "permission": "DELETE_PROJECT_API_TOKEN",
    "type": "project",
  },
  {
    "display_name": "Update variants",
    "permission": "UPDATE_FEATURE_ENVIRONMENT_VARIANTS",
    "type": "environment",
  },
  {
    "display_name": "Enable/disable toggles",
    "permission": "UPDATE_FEATURE_ENVIRONMENT",
    "type": "environment",
  },
  {
    "display_name": "Apply change requests",
    "permission": "APPLY_CHANGE_REQUEST",
    "type": "environment",
  },
  {
    "display_name": "Update CLIENT API tokens",
    "permission": "UPDATE_CLIENT_API_TOKEN",
    "type": "root",
  },
  {
    "display_name": "Create/edit project segment",
    "permission": "UPDATE_PROJECT_SEGMENT",
    "type": "project",
  },
  {
    "display_name": "Skip change request process",
    "permission": "SKIP_CHANGE_REQUEST",
    "type": "environment",
  },
  {
    "display_name": "Delete CLIENT API tokens",
    "permission": "DELETE_CLIENT_API_TOKEN",
    "type": "root",
  },
  {
    "display_name": "Read CLIENT API tokens",
    "permission": "READ_CLIENT_API_TOKEN",
    "type": "root",
  },
  {
    "display_name": "Approve/Reject change requests",
    "permission": "APPROVE_CHANGE_REQUEST",
    "type": "environment",
  },
  {
    "display_name": "Create FRONTEND API tokens",
    "permission": "CREATE_FRONTEND_API_TOKEN",
    "type": "root",
  },
  {
    "display_name": "Update FRONTEND API tokens",
    "permission": "UPDATE_FRONTEND_API_TOKEN",
    "type": "root",
  },
  {
    "display_name": "Delete FRONTEND API tokens",
    "permission": "DELETE_FRONTEND_API_TOKEN",
    "type": "root",
  },
  {
    "display_name": "Read FRONTEND API tokens",
    "permission": "READ_FRONTEND_API_TOKEN",
    "type": "root",
  },
  {
    "display_name": "Update feature dependency",
    "permission": "UPDATE_FEATURE_DEPENDENCY",
    "type": "project",
  },
  {
    "display_name": "Create tag types",
    "permission": "CREATE_TAG_TYPE",
    "type": "root",
  },
]
`;

exports[`Favor permission name over id migration correctly assigns permissions by name 2`] = `
[
  {
    "environment": null,
    "permission": "ADMIN",
    "role_id": 6,
  },
  {
    "environment": null,
    "permission": "CREATE_FEATURE",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "CREATE_STRATEGY",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "CREATE_ADDON",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "DELETE_ADDON",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "UPDATE_ADDON",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "UPDATE_FEATURE",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "DELETE_FEATURE",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "UPDATE_APPLICATION",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "UPDATE_TAG_TYPE",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "DELETE_TAG_TYPE",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "CREATE_PROJECT",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "UPDATE_PROJECT",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "DELETE_PROJECT",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "UPDATE_STRATEGY",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "DELETE_STRATEGY",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "UPDATE_CONTEXT_FIELD",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "CREATE_CONTEXT_FIELD",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "DELETE_CONTEXT_FIELD",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "UPDATE_FEATURE_VARIANTS",
    "role_id": 7,
  },
  {
    "environment": "default",
    "permission": "CREATE_FEATURE_STRATEGY",
    "role_id": 7,
  },
  {
    "environment": "development",
    "permission": "CREATE_FEATURE_STRATEGY",
    "role_id": 7,
  },
  {
    "environment": "production",
    "permission": "CREATE_FEATURE_STRATEGY",
    "role_id": 7,
  },
  {
    "environment": "default",
    "permission": "UPDATE_FEATURE_STRATEGY",
    "role_id": 7,
  },
  {
    "environment": "development",
    "permission": "UPDATE_FEATURE_STRATEGY",
    "role_id": 7,
  },
  {
    "environment": "production",
    "permission": "UPDATE_FEATURE_STRATEGY",
    "role_id": 7,
  },
  {
    "environment": "default",
    "permission": "DELETE_FEATURE_STRATEGY",
    "role_id": 7,
  },
  {
    "environment": "development",
    "permission": "DELETE_FEATURE_STRATEGY",
    "role_id": 7,
  },
  {
    "environment": "production",
    "permission": "DELETE_FEATURE_STRATEGY",
    "role_id": 7,
  },
  {
    "environment": "default",
    "permission": "UPDATE_FEATURE_ENVIRONMENT",
    "role_id": 7,
  },
  {
    "environment": "development",
    "permission": "UPDATE_FEATURE_ENVIRONMENT",
    "role_id": 7,
  },
  {
    "environment": "production",
    "permission": "UPDATE_FEATURE_ENVIRONMENT",
    "role_id": 7,
  },
  {
    "environment": "default",
    "permission": "UPDATE_FEATURE_ENVIRONMENT_VARIANTS",
    "role_id": 7,
  },
  {
    "environment": "development",
    "permission": "UPDATE_FEATURE_ENVIRONMENT_VARIANTS",
    "role_id": 7,
  },
  {
    "environment": "production",
    "permission": "UPDATE_FEATURE_ENVIRONMENT_VARIANTS",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "MOVE_FEATURE_TOGGLE",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "CREATE_SEGMENT",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "UPDATE_SEGMENT",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "DELETE_SEGMENT",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "READ_PROJECT_API_TOKEN",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "CREATE_PROJECT_API_TOKEN",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "DELETE_PROJECT_API_TOKEN",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "READ_CLIENT_API_TOKEN",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "READ_FRONTEND_API_TOKEN",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "CREATE_TAG_TYPE",
    "role_id": 7,
  },
  {
    "environment": null,
    "permission": "CREATE_FEATURE",
    "role_id": 9,
  },
  {
    "environment": null,
    "permission": "UPDATE_FEATURE",
    "role_id": 9,
  },
  {
    "environment": null,
    "permission": "DELETE_FEATURE",
    "role_id": 9,
  },
  {
    "environment": null,
    "permission": "UPDATE_PROJECT",
    "role_id": 9,
  },
  {
    "environment": null,
    "permission": "DELETE_PROJECT",
    "role_id": 9,
  },
  {
    "environment": null,
    "permission": "UPDATE_FEATURE_VARIANTS",
    "role_id": 9,
  },
  {
    "environment": "default",
    "permission": "CREATE_FEATURE_STRATEGY",
    "role_id": 9,
  },
  {
    "environment": "development",
    "permission": "CREATE_FEATURE_STRATEGY",
    "role_id": 9,
  },
  {
    "environment": "production",
    "permission": "CREATE_FEATURE_STRATEGY",
    "role_id": 9,
  },
  {
    "environment": "default",
    "permission": "UPDATE_FEATURE_STRATEGY",
    "role_id": 9,
  },
  {
    "environment": "development",
    "permission": "UPDATE_FEATURE_STRATEGY",
    "role_id": 9,
  },
  {
    "environment": "production",
    "permission": "UPDATE_FEATURE_STRATEGY",
    "role_id": 9,
  },
  {
    "environment": "default",
    "permission": "DELETE_FEATURE_STRATEGY",
    "role_id": 9,
  },
  {
    "environment": "development",
    "permission": "DELETE_FEATURE_STRATEGY",
    "role_id": 9,
  },
  {
    "environment": "production",
    "permission": "DELETE_FEATURE_STRATEGY",
    "role_id": 9,
  },
  {
    "environment": "default",
    "permission": "UPDATE_FEATURE_ENVIRONMENT",
    "role_id": 9,
  },
  {
    "environment": "development",
    "permission": "UPDATE_FEATURE_ENVIRONMENT",
    "role_id": 9,
  },
  {
    "environment": "production",
    "permission": "UPDATE_FEATURE_ENVIRONMENT",
    "role_id": 9,
  },
  {
    "environment": "default",
    "permission": "UPDATE_FEATURE_ENVIRONMENT_VARIANTS",
    "role_id": 9,
  },
  {
    "environment": "development",
    "permission": "UPDATE_FEATURE_ENVIRONMENT_VARIANTS",
    "role_id": 9,
  },
  {
    "environment": "production",
    "permission": "UPDATE_FEATURE_ENVIRONMENT_VARIANTS",
    "role_id": 9,
  },
  {
    "environment": null,
    "permission": "MOVE_FEATURE_TOGGLE",
    "role_id": 9,
  },
  {
    "environment": null,
    "permission": "READ_PROJECT_API_TOKEN",
    "role_id": 9,
  },
  {
    "environment": null,
    "permission": "CREATE_PROJECT_API_TOKEN",
    "role_id": 9,
  },
  {
    "environment": null,
    "permission": "DELETE_PROJECT_API_TOKEN",
    "role_id": 9,
  },
  {
    "environment": null,
    "permission": "UPDATE_FEATURE_DEPENDENCY",
    "role_id": 9,
  },
  {
    "environment": null,
    "permission": "CREATE_FEATURE",
    "role_id": 10,
  },
  {
    "environment": null,
    "permission": "UPDATE_FEATURE",
    "role_id": 10,
  },
  {
    "environment": null,
    "permission": "DELETE_FEATURE",
    "role_id": 10,
  },
  {
    "environment": null,
    "permission": "UPDATE_FEATURE_VARIANTS",
    "role_id": 10,
  },
  {
    "environment": "default",
    "permission": "CREATE_FEATURE_STRATEGY",
    "role_id": 10,
  },
  {
    "environment": "development",
    "permission": "CREATE_FEATURE_STRATEGY",
    "role_id": 10,
  },
  {
    "environment": "production",
    "permission": "CREATE_FEATURE_STRATEGY",
    "role_id": 10,
  },
  {
    "environment": "default",
    "permission": "UPDATE_FEATURE_STRATEGY",
    "role_id": 10,
  },
  {
    "environment": "development",
    "permission": "UPDATE_FEATURE_STRATEGY",
    "role_id": 10,
  },
  {
    "environment": "production",
    "permission": "UPDATE_FEATURE_STRATEGY",
    "role_id": 10,
  },
  {
    "environment": "default",
    "permission": "DELETE_FEATURE_STRATEGY",
    "role_id": 10,
  },
  {
    "environment": "development",
    "permission": "DELETE_FEATURE_STRATEGY",
    "role_id": 10,
  },
  {
    "environment": "production",
    "permission": "DELETE_FEATURE_STRATEGY",
    "role_id": 10,
  },
  {
    "environment": "default",
    "permission": "UPDATE_FEATURE_ENVIRONMENT",
    "role_id": 10,
  },
  {
    "environment": "development",
    "permission": "UPDATE_FEATURE_ENVIRONMENT",
    "role_id": 10,
  },
  {
    "environment": "production",
    "permission": "UPDATE_FEATURE_ENVIRONMENT",
    "role_id": 10,
  },
  {
    "environment": "default",
    "permission": "UPDATE_FEATURE_ENVIRONMENT_VARIANTS",
    "role_id": 10,
  },
  {
    "environment": "development",
    "permission": "UPDATE_FEATURE_ENVIRONMENT_VARIANTS",
    "role_id": 10,
  },
  {
    "environment": "production",
    "permission": "UPDATE_FEATURE_ENVIRONMENT_VARIANTS",
    "role_id": 10,
  },
  {
    "environment": null,
    "permission": "READ_PROJECT_API_TOKEN",
    "role_id": 10,
  },
  {
    "environment": null,
    "permission": "CREATE_PROJECT_API_TOKEN",
    "role_id": 10,
  },
  {
    "environment": null,
    "permission": "DELETE_PROJECT_API_TOKEN",
    "role_id": 10,
  },
  {
    "environment": null,
    "permission": "UPDATE_FEATURE_DEPENDENCY",
    "role_id": 10,
  },
]
`;
